### ISSUE TITLE
Use of "is True" throughout the code

### ISSUE THREAD

**Issue by NeilGirdhar** [networkx/networkx#3995]
Is the use of `if x is True` throughout the code intentional?  Testing _truthiness_ with `if x` is sufficient.

`x is True` is specifically [proscribed by PEP8](https://www.python.org/dev/peps/pep-0008/).  It means that calling a method that accepts a bool with the parameter value `1` produces falsy behavior.  This might lead to unexpected and confusing results—especially considering that `1 == True`!

**Comment by NeilGirdhar** [comment]
What do you expect this to do?
```python
d = DiGraph()
d.to_undirected(1)
```
or this?
```python
d.to_undirected(as_view=1)
```

You don't think it's confusing that passing in `1` is a synonym for `False`?

**Comment by dschult** [comment]
Why would anyone do that? 
I don't think that's a confusion we've ever had reported. People don't seem to have that trouble.  True/False are simple enough as they are. No need to use 1/0.

**Comment by NeilGirdhar** [comment]
Yes, I realize it's rare.  No one's going to report it when they can work around it after debugging. 
  As for when it can come up, someone might produce an int from checking if a number is even for example: `some_value & 1`.   Another way this fails is with alternate Boolean types like that found in `numpy`:
```
a = np.arange(10) < 3
print(a[0])  # True
assert bool(a[0])  # Yes.
assert a[0] is True  # Nope.
```

There's a reason that PEP8 warns that `if a is True` is even worse than `if a == True`.  I think it's bad, but I'll let you decide what you want to do.

**Comment by jond01** [comment]
Hi, pardon the interruption but I support @NeilGirdhar's side here...
A general guideline in python programming is duck typing, therefore checking for `x is True` is too strict. See the examples just above:
https://www.python.org/dev/peps/pep-0008/#function-annotations
However, refactoring this throughout the project may not be a top priority task

**Comment by dschult** [comment]
I agree with both of you that there are many places we currently have ```if x is True:``` where we should have ```if x```.  I suspect most (probably all) are cases where we never expect this input to be a result of a calculation, but rather a flag input value to indicate what the function should do.  

It is likely this sloppy treatment stemmed from getting used to using ```if x is None``` to indicate whether an input was provided explicitly, but that's no excuse for the sloppiness.

I would still argue there are a few places where ```is True``` is an appropriate and effective tool. One is for the ```edge_attr``` input to ```from_pandas_edgelist``` which can be str or int or iterable or True or None.  https://github.com/networkx/networkx/blob/8b0e2321d556daacfa0aaa140ef5789c712ce4ab/networkx/convert_matrix.py#L353
Perhaps there is a better way to arrange that API to avoid the need for ```if edge_attr is True```. Any ideas?

I cite [this webpage](https://docs.quantifiedcode.com/python-anti-patterns/readability/comparison_to_true.html) for pointing out that despite PEP8 saying ```if x is True``` is worse (without elaborating), that syntax is actually functionally important in some cases.

BUT....   NetworkX uses ```if x is True``` way more than needed and @NeilGirdhar's examples show a few cases where it isn't appropriate.  It should be cleaned up where appropriate.

**Comment by NeilGirdhar** [comment]
Thank you very much @dschult!

I guess I should come clean and explain my motivation for this issue.  I figured that even if you decided to postpone adding type annotations, eventually all Python libraries will have them.  To see how hard it would be, I ended up adding type annotations to graph and digraph.  I noticed a couple problems: the "is True" thing and the Liskov violation.  So I thought I would start issues to have those fixed so that if you end up adding the annotations, the road ahead is easy.  Sorry if this turned into a long discussion over a minor issue.

I understand your point about getting used to using `if x is None`.  After all `True` and `False` are singletons too.  The difference I think is that people make "truthy" and "falsy" objects, but no one makes "none-like" objects.

As for  your special case, I understand what you're trying to achieve.  Essentially `if edge_attr is True` is the same as saying `if type(edge_attr) is bool and edge_attr`.  This is one reason why linters and mypy complain for this kind of thing: they don't like type checking by identity: `if type(x) is SomeType`.  Instead, they want you to do: `if isinstance(edge_attr, bool)`.  I understand this is slightly longer.  The main benefits are that you can add the type annotations easily, and you avoid the weird behavior when people pass unexpected types in.

In defence of some of the coding patterns found in `networkx`, Python has had a mantra for a long time: "It's easier to ask for forgiveness than permission".   The issue is that type checkers will complain when you do this with types.  If a variable can have multiple types, then they want you to do something like this:
```
if isinstance(x, collections.abc.Container):
    # treat x as a container
elif isinstance(x, bool):
    # treat x as a bool
else:
    raise TypeError  # this error can come in really handy!
```

**Comment by dschult** [comment]
If we have to stop asking for forgiveness to appease the type-checkers, then the type checkers need to advance sufficiently to handle "asking for forgiveness rather than permission".  My understanding is that Code Design should not be affected by Python's type annotations. The annotations are supposed to help the programmers, not restrict them.

How hard would it be to enable mypy to allow:

```python
try:
    # treat x as a container
except:
    # treat x as a bool
```
Shouldn't it be able to tell that x can be a container or a bool?

**Comment by NeilGirdhar** [comment]
You're right.  Mypy knows that it can be either.  What it doesn't understand is that your try/except pattern regarding types.  I guess it could theoretically be made to work as you propose.  This is a philosophical question that's beyond me.  I wonder what Guido would say to this.

From a personal standpoint, I find the LBYL style simpler to reason about.  What happens if you have a variable type that's not a collection, but seems to act like it some networkx functions (that treat it as a container) , but not in others (that try to iterate over it)?  Then you will get very weird bugs since some of your networkx function treat it as a collection but others don't.

**Comment by mattall** [comment]
For what it's worth I went ahead and made a quick list of all the occurrences of `is True` `is False` and `is None` in the current main branch 40ae1b4. 

There are 495 results in 141 files.

If the `is None` pattern is always used deliberately to check whether an optional parameter was passed to a method then that leaves 60 results in 19 files for `is True` and `is False`. 

[where_is_is.txt](https://github.com/networkx/networkx/files/11580019/where_is_is.txt)

[where_is_is_True_is_False.txt](https://github.com/networkx/networkx/files/11580104/where_is_is_True_is_False.txt)

Would it be helpful if I started to make a few small pull requests addressing these? 

I'm only hesitant because I don't want to quash intentionally uses of the `is True` pattern as described in https://github.com/networkx/networkx/issues/3995#issuecomment-645567633.
